vm/devices/storage/disk_nvme/nvme_driver/src/queue_pair.rs

```diff
@@ -110,40 +111,133 @@ impl PendingCommands {
         );
         command.respond
     }
+
+    /// Save pending commands into a buffer.
+    pub fn save(&self) -> PendingCommandsSavedState {
+        let mut commands = Vec::new();
+        // Convert Slab into Vec.
+        for cmd in &self.commands {
+            commands.push(cmd.1.command);
+        }
+        PendingCommandsSavedState {
+            commands,
+            next_cid_high_bits: self.next_cid_high_bits.0,
+        }
+    }
+
+    /// Restore pending commands from the saved state.
+    pub fn restore(&mut self, saved_state: &PendingCommandsSavedState) -> anyhow::Result<()> {
+        let mut commands: Vec<(usize, PendingCommand)> = Vec::new();
+        for cmd in &saved_state.commands {
+            let (send, mut _recv) = mesh::oneshot::<nvme_spec::Completion>();
+            let pending_command = PendingCommand {
+                command: *cmd,
+                respond: send,
+            };
+            // Remove high CID bits to be used as a key.
+            let cid = cmd.cdw0.cid() & Self::CID_KEY_MASK;
+            commands.push((cid as usize, pending_command));
```

Comment:
Use .collect() for this path, too. Then you don't need to first make a vec and then make a slab.

Reply:
Converted to a single operator. Done.